app/models/targeting.rb

```diff
@@ -0,0 +1,29 @@
+class Targeting < ActiveRecord::Base
+
+  belongs_to :user
+  belongs_to :bookmark
+
+  has_many :targeting_hosts, :dependent => :destroy
+  has_many :hosts, :through => :targeting_hosts
+  has_many :template_invocations, :dependent => :destroy
+
+  validate :bookmark_or_query
+
+  def resolve_hosts!
+    fail _("Cannot resolve hosts without a user") if user.nil?
```

Comment:
small nitpick, we should try to use single quotes for strings if we don't do any interpolation